File: Leetcode/two_pointers/16_3SumClosest/Solution.py
class Solution(object):
    def threeSumClosest(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        nums.sort()
        res = nums[0]+nums[1]+nums[-1]
        for i in range(len(nums)-2):
            start, end = i+1, len(nums)-1
            while start<end:
                s = nums[i]+nums[start]+nums[end]
                if(abs(s-target)<abs(res-target)):
                    res = s
                if s>target:
                    end-=1
                else:
                    start+=1
        return res

File: Leetcode/two_pointers/16_3SumClosest/Solution.java
class Solution {
    public int threeSumClosest(int[] nums, int target) {
        Arrays.sort(nums);
        int res = nums[0]+nums[1]+nums[nums.length-1];
        for(int i = 0; i<nums.length-2; ++i){
            int start = i+1, end = nums.length-1;
            while(start < end){
                int sum = nums[i]+nums[start]+nums[end];
                if( Math.abs(sum-target)<Math.abs(res-target) ){
                    res = sum;
                }
                if(sum>target){ end-=1; }
                else{ start+=1; }
            }
        }
        return res;
    }
}
